### ISSUE TITLE
Feature request: tune_fit()

### ISSUE THREAD

**Issue by markfairbanks** [tidymodels/tune#156]
Are there any plans for a function like `tune_fit()` that tunes the hyperparameters and fits to the full dataset all in one step? See below.

```r
rf_model <- rand_forest(mtry = tune(), min_n = tune()) %>%
  set_mode("regression") %>%
  set_engine("ranger") %>%
  tune_fit(Sepal.Length ~ .,
           resamples = vfold_cv(iris_train, 5),
           grid = expand_grid(mtry = c(2,3),
                              min_n = c(20,30)),
           fit_data = iris_train)

rf_pred <- predict(rf_model, iris_test)
```

**Comment by BobMuenchen** [comment]
That would be very helpful. So far that process seems quite a bit harder in tidymodels than in caret.

**Comment by markfairbanks** [comment]
Here's the general shell of the function:

``` r
library(tidyverse)
library(parsnip)
library(tune)
library(rsample)
library(yardstick)

tune_fit <- function(model_spec,
                     formula,
                     resamples,
                     ...,
                     param_info = NULL,
                     grid = 10,
                     tune_metric = mae,
                     control = control_grid(),
                     fit_data) {
  
  tune_metric <- enquo(tune_metric)
  
  tune_result <- tune_grid(formula,
                           model = model_spec,
                           resamples = resamples,
                           grid = grid,
                           metrics = metric_set(!!tune_metric),
                           control = control)
  
  best_params <- tune_result %>%
    select_best(expr_text(metric), maximize = FALSE)
  
  model <- model_spec %>%
    finalize_model(best_params) %>%
    fit(formula, data = fit_data)
  
  model
}

set.seed(123)
split <- initial_split(iris)
iris_train <- training(split)
iris_test <- testing(split)

rf_model <- rand_forest(mtry = tune(), min_n = tune()) %>%
  set_mode("regression") %>%
  set_engine("ranger", importance = "impurity", num.threads = 6) %>%
  tune_fit(Sepal.Length ~ .,
           resamples = vfold_cv(iris_train, 5),
           grid = expand_grid(mtry = c(2,3),
                              min_n = c(20,30)),
           fit_data = iris_train)
#> i Creating pre-processing data to finalize unknown parameter: mtry

rf_pred <- predict(rf_model, iris_test)
```

<sup>Created on 2020-01-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by topepo** [comment]
`last_fit()` does a lot of this.

I'm extremely hesitant to automate things to this degree.  With this function, you never get the chance to evaluate whether the tuning values were appropriate. 

I have been debating having very high-level wrappers for specific models (think `tune_glmnet()`, `tune_xgboost()`) with the minimum recommended pre-processing built-in. They would not go near the test set though. 

I had to be talked into `last_fit()`. It is worthwhile but people will end up evaluating the test set repeatedly; we shouldn't enable that (`caret` doesn't even go that far). 

It might be worth adding a `fit` option to the `select_*` functions that would return a fitted model instead of a tibble of results.

**Comment by markfairbanks** [comment]
Interesting, I didn't know about `last_fit()`. Question though - doesn't this `caret` code replicate `tune_fit()`? Is the idea to move away from being able to do this?

```r
library(caret)

rf_model <- train(Sepal.Length ~ .,
                  data = iris_train,
                  method = "ranger",
                  tuneGrid = expand.grid(mtry = c(2,3),
                                         min.node.size = c(20,30),
                                         splitrule = "variance"),
                  trControl = trainControl(method = "cv", number = 5))
```

**Comment by topepo** [comment]
>  Is the idea to move away from being able to do this?

We are not trying to exactly replicate `caret`. I'd rather not fully automate every step here in an effort to make people think before they do; that is a major downside to using `caret` IMO. 

I'm okay without people having to write those 5 lines that `tune_fit()` automates. 

If `tune_fit()` does what you want, then why not just use that?

**Comment by markfairbanks** [comment]
> We are not trying to exactly replicate caret. I'd rather not fully automate every step here in an effort to make people think before they do; that is a major downside to using caret IMO.

Ah gotcha - makes perfect sense. Thanks for taking the time to respond!
